### ISSUE TITLE
Pylance fails with TypeError: Cannot read properties of undefined (reading 'nameWithScope')

### ISSUE THREAD

**Issue by dodancs** [microsoft/pyright#8509]
## Environment data

-   Pylance version: v2024.7.1 / also tried with v2024.7.100 (pre-release)
-   OS and version: Ubuntu 22.04.4 LTS
-   Python version (& distribution if applicable, e.g. Anaconda): 3.10.12 x64

## Code Snippet

This is just a part of the core, I do not want to share the whole, and also not sure where the issue is, so if you may require more of the code because it does not reproduce the issue just with this sample, I will provide more.

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import time

try:
    import requests
    from requests.auth import HTTPBasicAuth

    HAS_REQUESTS = True
except ImportError:
    HAS_REQUESTS = False


class PowerDNSError(Exception):
    def __init__(self, url, status_code, message):
        self.url = url
        self.status_code = status_code
        self.message = message
        super(PowerDNSError, self).__init__()


class PowerDNSClient:
    def __init__(self, host, port, prot, api_key, api_username, api_password, verify):
        self.url = "{prot}://{host}:{port}/api/v1".format(
            prot=prot, host=host, port=port
        )
        self.session = requests.Session()
        if api_key:
            self.session.headers.update({"X-API-Key": api_key})
        elif api_username and api_password:
            self.session.auth = HTTPBasicAuth(api_username, api_password)
        self.session.verify = verify

    def _handle_request(self, req):
        if req.status_code in [200, 201, 204]:
            if req.text:
                try:
                    return req.json()
                except Exception as e:
                    print(e)  # same as yield
            return dict()
        elif req.status_code == 404:
            error_message = "Not found"
        else:
            error_message = self._get_request_error_message(data=req)

        raise PowerDNSError(
            url=req.url, status_code=req.status_code, message=error_message
        )

    def _get_request_error_message(self, data):
        try:
            request_json = data.json()
            if "error" in request_json:
                request_error = request_json.get("error")
            elif "errors" in request_json:
                request_error = request_json.get("errors")
            else:
                request_error = "No error message found"
            return request_error
        except Exception:
            pass
        return data.text

    def _get_zones_url(self, server):
        return "{url}/servers/{server}/zones".format(url=self.url, server=server)

    def _get_zone_url(self, server, name):
        return "{url}/{name}".format(url=self._get_zones_url(server), name=name)

    def _get_search_url(self, server):
        return "{url}/servers/{server}/search-data".format(url=self.url, server=server)

    @staticmethod
    def _make_canonical(name):
        if not name.endswith("."):
            name += "."

        return name

    def get_zone(self, server, name):
        req = self.session.get(url=self._get_zone_url(server, name))
        if req.status_code == 422:  # zone does not exist
            return None
        return self._handle_request(req)

    def get_record(self, server, zone, name, rtype):
        """Search for a given record (name) in the specified zone."""
        url = self._get_search_url(server)
        params = {"q": name}

        resp = self._handle_request(self.session.get(url=url, params=params))

        # if returned an empty array, do not construct anything
        if not resp:
            return {}

        # Canonicalize record name and zone
        canonical_name = self._make_canonical(name)
        canonical_zone = self._make_canonical(zone)

        # Convert search result response to RRSet object
        rrset = dict(records=[], comments=[])

        for record in resp:
            if record["object_type"] != "record" or record["type"] != rtype:
                continue

            if record["name"] == canonical_name and record["zone"] == canonical_zone:

                if "name" not in rrset:
                    rrset["name"] = record["name"]
                if "type" not in rrset:
                    rrset["type"] = record["type"]
                if "ttl" not in rrset:
                    rrset["ttl"] = record["ttl"]

                rrentry = dict(content=record["content"], disabled=record["disabled"])
                rrset["records"].append(rrentry)

        return rrset

    def _get_request_data(
        self,
        changetype,
        server,
        zone,
        name,
        rtype,
        set_ptr=False,
        content=None,
        disabled=None,
        ttl=None,
    ):
        record_content = list()
        if content:
            for record in content:
                if rtype == "CNAME":
                    entry = dict(
                        content=self._make_canonical(record), disabled=disabled
                    )
                else:
                    entry = dict(content=record, disabled=disabled)
                if rtype in ["A", "AAAA"] and set_ptr:
                    entry["set-ptr"] = True

                record_content.append(entry)

        record = dict(
            name=name,
            type=rtype,
            changetype=changetype,
            records=record_content,
            ttl=ttl,
        )
        rrsets = list()
        rrsets.append(record)
        data = dict(rrsets=rrsets)
        return data

    def create_record(self, server, zone, name, rtype, content, disabled, ttl, set_ptr):
        url = self._get_zone_url(server=server, name=zone)

        # Ensure record name is fully canonical
        canonical_name = self._make_canonical(name)

        data = self._get_request_data(
            changetype="REPLACE",
            server=server,
            zone=zone,
            name=canonical_name,
            rtype=rtype,
            content=content,
            set_ptr=set_ptr,
            disabled=disabled,
            ttl=ttl,
        )
        req = self.session.patch(url=url, json=data)
        return self._handle_request(req)

    def delete_record(self, server, zone, name, rtype):
        canonical_name = self._make_canonical(name)
        url = self._get_zone_url(server=server, name=zone)
        data = self._get_request_data(
            changetype="DELETE",
            server=server,
            zone=zone,
            name=canonical_name,
            rtype=rtype,
        )
        req = self.session.patch(url=url, json=data)
        return self._handle_request(req)
```

## Repro Steps

1. Open the file
2. Maybe make some changes and save it, wait for Pylance to trigger
3. Find a squiggly underline at the first character of the file

![image](https://github.com/user-attachments/assets/ef8dcb61-6c7f-4954-b903-f1a70c1cc31c)

## Expected behavior

It works and does Pylance things.

## Actual behavior

It fails and does not lint.

## Logs

```
An internal error occurred while type checking file "/home/dominik/Nextcloud/Work/ansible/modules/ansible-module-powerdns/powerdns_records.py": TypeError: Cannot read properties of undefined (reading 'nameWithScope')
    at Object.e.getNameWithScope (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:1055888)
    at i.getTypeVar (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:993342)
    at i.setTupleTypeVar (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:993261)
    at /home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:996534
    at Array.forEach (<anonymous>)
    at o.setTupleTypeVar (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:996519)
    at L (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:952432)
    at P (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:947882)
    at U (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:951992)
    at U.next (<anonymous>)
    at V (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:950389)
    at V.next (<anonymous>)
    at z (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:950181)
    at Dt (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:647497)
    at He (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:617474)
    at Ze (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:618389)
    at o (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:728996)
    at /home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:729434
    at d (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:943122)
    at hn (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:729326)
    at /home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:622546
    at d (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:943122)
    at et (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:622293)
    at ya (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:790031)
    at wa (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:800376)
    at a.timeOperation (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:1210851)
    at /home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:191459
    at Sa (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:800807)
    at a.timeOperation (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:1210851)
    at D (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:191426)
    at /home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:192498
    at N (/home/dominik/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/pyright.bundle.js:1:191549)  Pylance
```

**Comment by rchiodo** [comment]
Transferring to Pyright. Core typechecker issue.

**Comment by erictraut** [comment]
There's not enough information here for me to diagnose, and I'm not able to repro the problem with the above code. If you can provide repro steps, I'd be happy to investigate further.

**Comment by Work-Git-Hub** [comment]
I've faced the exact same issue 

Pylance version: v2024.7.1 / also tried with v2024.7.100 (pre-release)
OS and version: Windows 10
Python version: 3.4 x64

I think issue has something to do with VSCode 

I started happening after the latest update 1.91.1


![image](https://github.com/user-attachments/assets/6dcf3fda-6927-4d23-adaa-9bbc1bc9053f)

Hopefully this helps. I can provide more steps as well if this doesn't suffice.

**Comment by erictraut** [comment]
@Work-Git-Hub, I need a way to repro the problem. Is it reproducible for you? If so, can you narrow it down to a particular piece of code and post that code sample here? Alternatively, if the code is in a public repo, please point me to the repo location.

**Comment by erictraut** [comment]
After digging deeper, I found a theoretical way for this error to repro, but it involves a corrupted definition of the `tuple` class in the `builtins.pyi` typeshed stub. That's pretty unlikely, so I'm guessing there's some other condition that I haven't yet uncovered. In any event, I've added some additional defensive code to detect the condition and prevent a crash. This will be addressed in the next release of pyright.

**Comment by dodancs** [comment]
Is it possible that other extensions are causing the issue? Or perhaps there is an issue because my system has a never version of Python, etc. that has introduced breaking changes?

I can try and boot up a fresh Ubuntu 24.04 desktop VM, install vscode and the plugin and try and use my code sample to see whether it can be reproduced on a virgin system with nothing but vscode and the extension.

**Comment by erictraut** [comment]
@dodancs, this wouldn't be caused by tools extensions. It would be caused by a particular combination of Python code (both the code in your program and code that it imports from libraries you have installed) that pyright is analyzing.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.374.
